client/components/data/domain-management/nameservers/index.jsx

```diff
@@ -1,80 +1,93 @@
 ];
 
 function getStateFromStores( props ) {
-	let domains;
-
-	if ( props.selectedSite ) {
-		domains = DomainsStore.getBySite( props.selectedSite.ID );
-	}
-
 	return {
-		domains,
+		domains: props.domains,
 		nameservers: NameserversStore.getByDomainName( props.selectedDomainName ),
 		selectedDomainName: props.selectedDomainName,
-		selectedSite: props.selectedSite
+		// undefined is not permitted by the passed component
+		// so we have to explicitly choose a bool in that case
+		selectedSite: props.selectedSite || false,
 	};
 }
 
-const NameserversData = React.createClass( {
-	propTypes: {
-		component: React.PropTypes.func.isRequired,
-		selectedDomainName: React.PropTypes.string.isRequired,
-		sites: React.PropTypes.object.isRequired
-	},
+export class NameserversData extends Component {
+	static propTypes = {
+		component: PropTypes.func.isRequired,
+		selectedDomainName: PropTypes.string.isRequired,
+		selectedSite: PropTypes.object,
+	};
 
-	mixins: [ observe( 'sites' ) ],
+	constructor( props ) {
+		super( props );
 
-	componentWillMount() {
-		this.loadDomains();
 		this.loadNameservers();
-	},
+	}
 
 	componentWillUpdate() {
-		this.loadDomains();
 		this.loadNameservers();
-	},
-
-	loadDomains() {
-		const selectedSite = this.props.sites.getSelectedSite();
-
-		if ( this.prevSelectedSite !== selectedSite ) {
-			upgradesActions.fetchDomains( selectedSite.ID );
-
-			this.prevSelectedSite = selectedSite;
-		}
-	},
+	}
 
-	loadNameservers() {
-		upgradesActions.fetchNameservers( this.props.selectedDomainName );
-	},
+	loadNameservers = () => fetchNameservers( this.props.selectedDomainName );
 
 	render() {
+		const { selectedSite } = this.props;
+
 		return (
-			<StoreConnection
-				component={ this.props.component }
-				stores={ stores }
-				getStateFromStores={ getStateFromStores }
-				selectedDomainName={ this.props.selectedDomainName }
-				selectedSite={ this.props.sites.getSelectedSite() } />
+			<div>
+				<QuerySiteDomains siteId={ selectedSite && selectedSite.ID } />
+				<StoreConnection
+					component={ this.props.component }
+					domains={ this.props.domains }
+					stores={ stores }
+					getStateFromStores={ getStateFromStores }
+					selectedDomainName={ this.props.selectedDomainName }
+					selectedSite={ selectedSite }
+				/>
+			</div>
 		);
 	}
-} );
+}
+
+const mapStateToProps = state => {
+	const selectedSite = getSelectedSite( state );
+	const domains = getDomainsBySite( state, selectedSite );
+
+	return {
+		domains: {
+			// this doesn't appear to be currently
+			// stored in our app state as it was in
+			// the domains store. we know if we are
+			// currently fetching, but not if a fetch
+			// has taken place. in this case, I'm
+			// using the length of the array to indicate
+			// `hasLoaded` because this component should
+			// never exist in the app without at least
+			// one domain. thus, if we have at least one
+			// then we have loaded them.
+			hasLoadedFromServer: !! domains.length,
```

Comment:
@gwwar sigh it's not so clear. the component that is nested is using `hasLoadedFromServer` for initial load and has no distinction between _loaded_ and _loading_. That is to say that I don't think we can make this small change without making lots of changes to the nested component.

As it is, the data may be stale, but new data should pop in once it loads. There will be that initial swap of data, but on the other hand I think we'll get placeholders instead of potentially stale data each time the endpoint gets polled.

Are you comfortable leaving this as-is for now or do you think we should just add `hasLoadedFromServer` into the state tree?